drivers/input/input_pmw3610.c

```diff
@@ -0,0 +1,433 @@
+#define REST1_DOWNSHIFT_INIT 0x0f
+
+#define PRODUCT_ID_PMW3610 0x3e
+#define SPI_WRITE BIT(7)
+#define MOTION_STATUS_MOTION BIT(7)
+#define SPI_CLOCK_ON_REQ_ON 0xba
+#define SPI_CLOCK_ON_REQ_OFF 0xb5
+#define RES_STEP_INV_X_BIT 6
+#define RES_STEP_INV_Y_BIT 5
+#define POWER_UP_WAKEUP 0x96
+#define SHUTDOWN_ENABLE 0xe7
+#define SPI_PAGE0_1 0xff
+#define SPI_PAGE1_0 0x00
+
+#define PMW3610_DATA_SIZE_BITS 12
+
+#define RESET_DELAY_MS 10
+#define INIT_OBSERVATION_DELAY_MS 10
+#define CLOCK_ON_DELAY_US 300
+
+struct pmw3610_config {
+	struct spi_dt_spec spi;
+	struct gpio_dt_spec motion_gpio;
+	struct gpio_dt_spec reset_gpio;
+	uint16_t axis_x;
+	uint16_t axis_y;
+	bool invert_x;
+	bool invert_y;
+};
+
+struct pmw3610_data {
+	const struct device *dev;
+	struct k_work motion_work;
+	struct gpio_callback motion_cb;
+};
+
+static int pmw3610_read(const struct device *dev,
+			uint8_t addr, uint8_t *value, uint8_t len)
+{
+	const struct pmw3610_config *cfg = dev->config;
+
+	const struct spi_buf tx_buf = {
+		.buf = &addr,
+		.len = sizeof(addr),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	struct spi_buf rx_buf[] = {
+		{
+			.buf = NULL,
+			.len = sizeof(addr),
+		},
+		{
+			.buf = value,
+			.len = len,
+		},
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	return spi_transceive_dt(&cfg->spi, &tx, &rx);
+}
+
+static int pmw3610_read_reg(const struct device *dev, uint8_t addr, uint8_t *value)
+{
+	return pmw3610_read(dev, addr, value, 1);
+}
+
+static int pmw3610_write_reg(const struct device *dev, uint8_t addr, uint8_t value)
+{
+	const struct pmw3610_config *cfg = dev->config;
+
+	uint8_t write_buf[] = {addr | SPI_WRITE, value};
+	const struct spi_buf tx_buf = {
+		.buf = write_buf,
+		.len = sizeof(write_buf),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	return spi_write_dt(&cfg->spi, &tx);
+}
+
+static int pmw3610_spi_clk_on(const struct device *dev)
+{
+	int ret;
+
+	ret = pmw3610_write_reg(dev, PMW3610_SPI_CLK_ON_REQ, SPI_CLOCK_ON_REQ_ON);
+
+	k_sleep(K_USEC(CLOCK_ON_DELAY_US));
+
+	return ret;
+}
+
+static int pmw3610_spi_clk_off(const struct device *dev)
+{
+	return pmw3610_write_reg(dev, PMW3610_SPI_CLK_ON_REQ, SPI_CLOCK_ON_REQ_OFF);
+}
+
+static void pmw3610_motion_work_handler(struct k_work *work)
+{
+	struct pmw3610_data *data = CONTAINER_OF(
+			work, struct pmw3610_data, motion_work);
+	const struct device *dev = data->dev;
+	const struct pmw3610_config *cfg = dev->config;
+	uint8_t burst_data[4];
+	int32_t x, y;
+	int ret;
+
+	ret = pmw3610_read(dev, PMW3610_BURST_READ, burst_data, sizeof(burst_data));
+	if (ret < 0) {
+		return;
+	}
+
+	if ((burst_data[BURST_MOTION] & MOTION_STATUS_MOTION) == 0x00) {
+		return;
+	}
+
+	x = ((burst_data[BURST_DELTA_XY_H] << 4) & 0xf00) | burst_data[BURST_DELTA_X_L];
+	y = ((burst_data[BURST_DELTA_XY_H] << 8) & 0xf00) | burst_data[BURST_DELTA_Y_L];
```

Comment:
question: same mask for both cases, is this correct?

Reply:
yeah the mask is after the shift and the shift is different, I'm somewhat surprised that the shift does not push the bits out of the uint8_t but it seems to work, tried

```c
        uint8_t burst_data[4];
        int32_t x, y;
        burst_data[0] = 0x12;
        burst_data[1] = 0x03;
        burst_data[2] = 0x04;
        x = ((burst_data[0] << 4) & 0xf00) | burst_data[1];
        y = ((burst_data[0] << 8) & 0xf00) | burst_data[2];
        printf("%04x %04x\n", x, y);
```

```
0103 0204
```

Reply:
burst_data[0] is implicitly promoted to an int, so this is guaranteed to work.

Reply:
Gotta love this language, always learning something. :-)

Reply:
actually, should'nt you use `U` suffixes for the constants? IIRC binary ops are undefined on signed integers

Reply:
is that the case? that's usually something I don't pay too much attention to but I can change it if it does